### PR TITLE
GH-39704: [C++][Parquet] Benchmark levels decoding 

### LINKED ISSUE #39704: [C++][Parquet] Benchmark Level Decoding
### Describe the enhancement requested

Parquet Level Decoding requires more benchmarks

### Component(s)

Benchmarking, C++, Parquet

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

This patch add the level-decoding benchmark. It test:
1. Different max-level (for flat type, maximum level would be 1, for nested type, it would grows)
2. With different repeat ( repeated null / non-null is different from non-repeated data)
3. With different read-batch size. This part of logic is a bit tricky in original code

### What changes are included in this PR?

Add Level decoding benchmark

### Are these changes tested?

No need

### Are there any user-facing changes?

no

* Closes: #39704

### CODE REVIEW

**Comment by mapleFU** [comment]
Benchmark:

```
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1          2729 ns         2722 ns       254899
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7          2776 ns         2731 ns       252060
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1024       2732 ns         2722 ns       257173
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1          2229 ns         2139 ns       348028
ReadLevels/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1          2084 ns         1985 ns       354815
ReadLevels/MaxLevel:3/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1          1777 ns         1741 ns       404353
ReadLevels/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7          2050 ns         1987 ns       352262
```

**Comment by mapleFU** [comment]
This benchmark shows that, when not highly repeated, the RLE without bitpacking is slow 😅

**Comment by alippai** [comment]
Is this use case relevant here? https://github.com/apache/arrow/issues/34510

Reading a non nullable fixed size list is missing the fast path, it’d nice to see it in the benchmark (even if not improving yet). With all the AI nowadays I assume tensor storage will be more and more common.

**Comment by mapleFU** [comment]
> Reading a non nullable fixed size list is missing the fast path

Yeah I think it's related, I think I can optimize unpack later, but maybe I need some help in optimizing RLE

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 1]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
     ->Args({2, 1000, true})
     ->Args({2, 1000, false});
 
+void GenerateLevels(int level_repeats, int max_level, int num_levels,
+                    std::vector<int16_t>& input_levels) {
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:223]: Let's please avoid mutable references. Also, why is it named `input_levels`?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 2]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+  // Generate random levels
+  std::default_random_engine gen(/*seed=*/1943);
+  std::uniform_int_distribution<int16_t> d(0, max_level);
+  for (int i = 0; i < num_levels;) {
+    int16_t current_level = d(gen);  // level repeat `level_repeats` times
+    for (int j = 0; j < level_repeats; ++j) {
+      input_levels.push_back(current_level);
+      ++i;
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:231]: What if `i >= num_levels`?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 3]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+    }
+  }
+}
+
+void EncodeLevels(Encoding::type encoding, int16_t max_level, int num_levels,
+                  const int16_t* input_levels, std::vector<uint8_t>& bytes) {
+  LevelEncoder encoder;
+  int levels_count = 0;
+  bytes.resize(2 * num_levels);
+  ASSERT_EQ(2 * num_levels, static_cast<int>(bytes.size()));
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:241]: Is this useful?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 4]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+    int16_t current_level = d(gen);  // level repeat `level_repeats` times
+    for (int j = 0; j < level_repeats; ++j) {
+      input_levels.push_back(current_level);
+      ++i;
+    }
+  }
+}
+
+void EncodeLevels(Encoding::type encoding, int16_t max_level, int num_levels,
+                  const int16_t* input_levels, std::vector<uint8_t>& bytes) {
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:237]: Can we please avoid mutable refs?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 5]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+  LevelEncoder encoder;
+  int levels_count = 0;
+  bytes.resize(2 * num_levels);
+  ASSERT_EQ(2 * num_levels, static_cast<int>(bytes.size()));
+  // encode levels
+  if (encoding == Encoding::RLE) {
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:243]: Is there a case where we don't use RLE for levels?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 6]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+
+static void DecodeLevels(Encoding::type level_encoding, int16_t max_level, int num_levels,
+                         int batch_size, int level_repeat_count,
+                         ::benchmark::State& state) {
+  std::vector<uint8_t> bytes;
+  {
+    std::vector<int16_t> input_levels;
+    GenerateLevels(/*level_repeats=*/level_repeat_count, /*max_repeat_factor=*/max_level,
+                   num_levels, input_levels);
+    // Print generated levels
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:266]: This comment doesn't seem accurate?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 7]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+    decoder.SetData(level_encoding, max_level, num_levels, bytes.data(),
+                    static_cast<int>(bytes.size()));
+    // Decode multiple times with batch_size
+    while (true) {
+      int levels_decoded = decoder.Decode(batch_size, output_levels.data());
+      if (levels_decoded == 0) {
+        break;
+      }
+    }
+  }
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:282]: Can you call `SetItemsProcessed` at the end? We should make sure benchmark numbers are intelligible.

**Review by pitrou — CHANGES_REQUESTED** [review]
Good idea @mapleFU . Please see my comments below.

File: cpp/src/parquet/column_reader_benchmark.cc (hunk 5 above)
Comment (mapleFU) [reply cpp/src/parquet/column_reader_benchmark.cc:243]: Let me add benchmark for BITPACK

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 8]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
 
+void GenerateLevels(int level_repeats, int max_level, int num_levels,
+                    std::vector<int16_t>& input_levels) {
+  // Generate random levels
+  std::default_random_engine gen(/*seed=*/1943);
+  std::uniform_int_distribution<int16_t> d(0, max_level);
+  for (int i = 0; i < num_levels;) {
+    int16_t current_level = d(gen);  // level repeat `level_repeats` times
+    for (int j = 0; j < level_repeats; ++j) {
+      input_levels.push_back(current_level);
```
Comment (emkornfield) [inline cpp/src/parquet/column_reader_benchmark.cc:230]: nit: you could use vector::insert to repeat the value for you.

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 9]
```diff
@@ -219,5 +219,87 @@ BENCHMARK(RecordReaderReadRecords)
...
+    GenerateLevels(/*level_repeats=*/level_repeat_count, /*max_repeat_factor=*/max_level,
+                   num_levels, input_levels);
+    // Print generated levels
+    EncodeLevels(level_encoding, max_level, num_levels, input_levels.data(), bytes);
+  }
+
+  LevelDecoder decoder;
+  std::vector<int16_t> output_levels(num_levels);
+  for (auto _ : state) {
+    decoder.SetData(level_encoding, max_level, num_levels, bytes.data(),
```
Comment (emkornfield) [inline cpp/src/parquet/column_reader_benchmark.cc:273]: just want to check that we want to include this in timing (I think it is cheap, but I guess a benchmark should show whether it is or isn't).

**Comment by mapleFU** [comment]
@emkornfield @pitrou Updated, so sorry for the delaying

**Comment by mapleFU** [comment]
Result in my MacOS with Release(-O2):

```
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1                  3122 ns         3123 ns       225916 bytes_per_second=4.8286G/s items_per_second=2.59233G/s
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7                  8642 ns         8640 ns        81130 bytes_per_second=1.74531G/s items_per_second=937.005M/s
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1024               1124 ns         1124 ns       617709 bytes_per_second=13.4136G/s items_per_second=7.20137G/s
ReadLevels/MaxLevel:1/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1                  2412 ns         2414 ns       290082 bytes_per_second=6.24778G/s items_per_second=3.35425G/s
ReadLevels/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1                  2403 ns         2401 ns       292828 bytes_per_second=6.27942G/s items_per_second=3.37124G/s
ReadLevels/MaxLevel:3/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1                  2201 ns         2202 ns       320742 bytes_per_second=6.84694G/s items_per_second=3.67592G/s
ReadLevels/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7                  7836 ns         7829 ns        90506 bytes_per_second=1.92618G/s items_per_second=1034.11M/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1          1724 ns         1726 ns       397519 bytes_per_second=8.73636G/s items_per_second=4.6903G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7          1694 ns         1695 ns       411450 bytes_per_second=8.89878G/s items_per_second=4.7775G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1024       1694 ns         1695 ns       409966 bytes_per_second=8.89797G/s items_per_second=4.77706G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1          1668 ns         1669 ns       414886 bytes_per_second=9.03668G/s items_per_second=4.85153G/s
ReadLevels_BitPack/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1          1760 ns         1761 ns       395125 bytes_per_second=8.56206G/s items_per_second=4.59672G/s
ReadLevels_BitPack/MaxLevel:3/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1          1733 ns         1734 ns       402188 bytes_per_second=8.69522G/s items_per_second=4.66821G/s
ReadLevels_BitPack/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7          1754 ns         1756 ns       396380 bytes_per_second=8.58685G/s items_per_second=4.61003G/s
```

**Review by wgtmac — APPROVED** [review]

**Comment by mapleFU** [comment]
Ping @pitrou for help

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 10]
```diff
@@ -219,5 +219,99 @@ BENCHMARK(RecordReaderReadRecords)
...
+
+void EncodeLevels(Encoding::type encoding, int16_t max_level, int num_levels,
+                  const int16_t* input_levels, std::vector<uint8_t>* bytes) {
+  LevelEncoder encoder;
+  bytes->resize(2 * num_levels);
+  // encode levels
+  if (encoding == Encoding::RLE) {
+    // leave space to write the rle length value
+    encoder.Init(encoding, max_level, num_levels, bytes->data() + sizeof(int32_t),
+                 static_cast<int>(bytes->size()));
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:243]: Will this write past the end of `bytes`? Do you need to change the `bytes->resize` call for this?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 11]
```diff
@@ -219,5 +219,99 @@ BENCHMARK(RecordReaderReadRecords)
...
+      if (levels_decoded == 0) {
+        break;
+      }
+    }
+  }
+  state.SetBytesProcessed(state.iterations() * num_levels * sizeof(int16_t));
+  state.SetItemsProcessed(state.iterations() * num_levels);
+}
+
+static void ReadLevels(::benchmark::State& state) {
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:284]: Can you rename this `ReadLevels_Rle` for consistency?

File: cpp/src/parquet/column_reader_benchmark.cc [hunk 12]
```diff
@@ -219,5 +219,99 @@ BENCHMARK(RecordReaderReadRecords)
...
+  std::vector<uint8_t> bytes;
+  {
+    std::vector<int16_t> input_levels;
+    GenerateLevels(/*level_repeats=*/level_repeat_count, /*max_repeat_factor=*/max_level,
+                   num_levels, &input_levels);
+    EncodeLevels(level_encoding, max_level, num_levels, input_levels.data(), &bytes);
+  }
+
+  LevelDecoder decoder;
+  std::vector<int16_t> output_levels(num_levels);
```
Comment (pitrou) [inline cpp/src/parquet/column_reader_benchmark.cc:266]: Nit: you probably just need `batch_size`, not `num_levels`. That's not really important, though.

**Comment by pitrou** [comment]
Strange phenomenon: we get results `bytes_per_second=11.7241G/s items_per_second=6.29431G/s`, where `bytes_per_second` is not equal to `2 * items_per_second`.

**Comment by pitrou** [comment]
Oh, it seems Google benchmark has a weird behavior here. Unrelated to this PR though.

**Comment by pitrou** [comment]
Posted https://github.com/google/benchmark/issues/1749 for the Google benchmark oddity.

File: cpp/src/parquet/column_reader_benchmark.cc (hunk 12 above)
Comment (mapleFU) [reply cpp/src/parquet/column_reader_benchmark.cc:266]: Hmmm during reading, isn't they're both important?

File: cpp/src/parquet/column_reader_benchmark.cc (hunk 10 above)
Comment (mapleFU) [reply cpp/src/parquet/column_reader_benchmark.cc]: Hmmm I use estimated size for it now

File: cpp/src/parquet/column_reader_benchmark.cc (hunk 12 above)
Comment (pitrou) [reply cpp/src/parquet/column_reader_benchmark.cc:270]: You're reading `batch_size` below.

**Comment by pitrou** [comment]
FTR, benchmark numbers here:
```
--------------------------------------------------------------------------------------------------------------------------------------------
Benchmark                                                                                  Time             CPU   Iterations UserCounters...
--------------------------------------------------------------------------------------------------------------------------------------------
ReadLevels_Rle/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1              2487 ns         2488 ns       288334 bytes_per_second=6.06116Gi/s items_per_second=3.25406G/s
ReadLevels_Rle/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7              8071 ns         8072 ns        86861 bytes_per_second=1.86809Gi/s items_per_second=1.00292G/s
ReadLevels_Rle/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1024            826 ns          828 ns       841216 bytes_per_second=18.2145Gi/s items_per_second=9.77881G/s
ReadLevels_Rle/MaxLevel:1/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1              2209 ns         2211 ns       314019 bytes_per_second=6.81903Gi/s items_per_second=3.66094G/s
ReadLevels_Rle/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1              2110 ns         2112 ns       331250 bytes_per_second=7.13955Gi/s items_per_second=3.83302G/s
ReadLevels_Rle/MaxLevel:3/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1              1904 ns         1906 ns       368359 bytes_per_second=7.91382Gi/s items_per_second=4.2487G/s
ReadLevels_Rle/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7              7673 ns         7675 ns        90873 bytes_per_second=1.96488Gi/s items_per_second=1.05489G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1          1338 ns         1342 ns       522900 bytes_per_second=11.2397Gi/s items_per_second=6.03429G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7          1342 ns         1345 ns       521761 bytes_per_second=11.2118Gi/s items_per_second=6.01927G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1024       1340 ns         1343 ns       520704 bytes_per_second=11.2263Gi/s items_per_second=6.02705G/s
ReadLevels_BitPack/MaxLevel:1/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1          1327 ns         1330 ns       526037 bytes_per_second=11.3356Gi/s items_per_second=6.08578G/s
ReadLevels_BitPack/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:1          1335 ns         1338 ns       520920 bytes_per_second=11.2687Gi/s items_per_second=6.04985G/s
ReadLevels_BitPack/MaxLevel:3/NumLevels:8096/BatchSize:2048/LevelRepeatCount:1          1327 ns         1329 ns       526969 bytes_per_second=11.3443Gi/s items_per_second=6.09043G/s
ReadLevels_BitPack/MaxLevel:3/NumLevels:8096/BatchSize:1024/LevelRepeatCount:7          1336 ns         1338 ns       522798 bytes_per_second=11.2696Gi/s items_per_second=6.05033G/s
```

File: cpp/src/parquet/column_reader_benchmark.cc (hunk 12 above)
Comment (mapleFU) [reply cpp/src/parquet/column_reader_benchmark.cc:270]:
```
    // Decode multiple times with batch_size
    while (true) {
      int levels_decoded = decoder.Decode(batch_size, output_levels.data());
      if (levels_decoded == 0) {
        break;
      }
    }
```

Emmm I will read `batch_size` until read finished

**Comment by mapleFU** [comment]
@pitrou I think the benchmark result shows "batch_size" should be take into consideration, for example, when the batchsize grows, BITPACK doesn't get improved, however, RLE is well optimized

File: cpp/src/parquet/column_reader_benchmark.cc (hunk 12 above)
Comment (pitrou) [reply cpp/src/parquet/column_reader_benchmark.cc:270]: This doesn't answer the comment, does it?

Comment (mapleFU) [reply cpp/src/parquet/column_reader_benchmark.cc:270]:
```c++
    while (true) {
      int levels_decoded = decoder.Decode(batch_size, output_levels.data());
      std::cout << levels_decoded << '\n';
      if (levels_decoded == 0) {
        break;
      }
    }
```

A sample output:

```
levels decoded: 1024
levels decoded: 1024
levels decoded: 1024
levels decoded: 1024
levels decoded: 1024
levels decoded: 1024
levels decoded: 1024
levels decoded: 928
levels decoded: 0
```

I've read the whole data here

Comment (pitrou) [reply cpp/src/parquet/column_reader_benchmark.cc:270]: Please re-read. I'm talking about `std::vector<int16_t> output_levels(num_levels)`, not about how many bytes you read. You're always reading at the start of `output_levels`.

**Comment by mapleFU** [comment]
Mybad, change vector for output to `batch_size`

**Pull request merged**